runtime/src/dice.rs

```diff
@@ -0,0 +1,68 @@
+// Licensed under the Apache-2.0 license
+
+use crate::RuntimeErr;
+use caliptra_drivers::{CaliptraResult, DataVault};
+use caliptra_x509::{Ecdsa384CertBuilder, Ecdsa384Signature, FmcAliasCertTbs, LocalDevIdCertTbs};
+
+extern "C" {
+    static mut LDEVID_TBS_ORG: u8;
+    static mut FMCALIAS_TBS_ORG: u8;
+}
+
+enum CertType {
+    LDevId,
+    FmcAlias,
+}
+
+/// Copy LDevID certificate produced by ROM to `cert` buffer
+///
+/// Returns the number of bytes written to `cert`
+pub fn copy_ldevid_cert(cert: &mut [u8]) -> CaliptraResult<usize> {
+    cert_from_dccm(cert, CertType::LDevId)
+}
+
+/// Copy FMC Alias certificate produced by ROM to `cert` buffer
+///
+/// Returns the number of bytes written to `cert`
+pub fn copy_fmc_alias_cert(cert: &mut [u8]) -> CaliptraResult<usize> {
+    cert_from_dccm(cert, CertType::FmcAlias)
+}
+
+/// Copy a certificate from `dccm_offset`, append signature, and write the
+/// output to `cert`.
+fn cert_from_dccm(cert: &mut [u8], cert_type: CertType) -> CaliptraResult<usize> {
+    let data_vault = DataVault::default();
+    let (tbs, sig) = match cert_type {
+        CertType::LDevId => {
+            let tbs = unsafe {
+                let ptr = &mut LDEVID_TBS_ORG as *mut u8;
+                core::slice::from_raw_parts_mut(ptr, LocalDevIdCertTbs::TBS_TEMPLATE_LEN)
```

Comment:
It makes me a _little_ bit nervous that there's no length field here being written by ROM. Perhaps we need a feature-request to write that out.

Reply:
Agreed this is a bit concerning. But I think there are competing considerations:

1. We could maybe read the value from the hand-off table. But an attacker which can control this value has a DCCM-read primitive. So from that perspective, a hard-coded value may be safer.

2. As you pointed out, the hard-coded value is a bit error prone. Especially if RT needs to work on multiple different ROM releases.

Reply:
Closing this comment, we can track the enhancement in https://github.com/chipsalliance/caliptra-sw/pull/308